Zend/zend_string.c

```diff
@@ -78,71 +88,93 @@ void zend_known_interned_strings_init(zend_string ***strings, uint32_t *count)
 {
-#ifndef ZTS
 	zend_string *str;
 
-	zend_hash_init(&CG(interned_strings), 1024, NULL, _str_dtor, 1);
+	if (ZEND_INTERNED_STRINGS_UNINITIALIZED != zend_strings_storage_type || !(storage_type > ZEND_INTERNED_STRINGS_UNINITIALIZED)) {
+		return;
+	}
+
+	zend_init_interned_strings_ht(&interned_strings_permanent);
+	zend_init_interned_strings_ht(&interned_strings_volatile);
 
-	CG(interned_strings).nTableMask = -CG(interned_strings).nTableSize;
-	HT_SET_DATA_ADDR(&CG(interned_strings), pemalloc(HT_SIZE(&CG(interned_strings)), 1));
-	HT_HASH_RESET(&CG(interned_strings));
-	CG(interned_strings).u.flags |= HASH_FLAG_INITIALIZED;
+	zend_new_interned_string = zend_new_interned_string_permanent_int;
+	zend_interned_strings_snapshot = zend_interned_strings_snapshot_int;
+	zend_interned_strings_restore = zend_interned_strings_restore_int;
 
 	/* interned empty string */
 	str = zend_string_alloc(sizeof("")-1, 1);
 	ZSTR_VAL(str)[0] = '\000';
-	CG(empty_string) = zend_new_interned_string_int(str);
-#endif
-
+	empty_string = zend_new_interned_string_permanent_int(str);
+#ifndef ZTS
 	/* one char strings (the actual interned strings are going to be created by ext/opcache) */
 	memset(CG(one_char_string), 0, sizeof(CG(one_char_string)));
+#endif
 
 	/* known strings */
 	zend_intern_known_strings(known_strings, (sizeof(known_strings) / sizeof(known_strings[0])) - 1);
+#ifndef ZTS
 	zend_known_interned_strings_init(&CG(known_strings), &CG(known_strings_count));
+#endif
 
-	zend_new_interned_string = zend_new_interned_string_int;
-	zend_interned_strings_snapshot = zend_interned_strings_snapshot_int;
-	zend_interned_strings_restore = zend_interned_strings_restore_int;
+	zend_strings_storage_type = storage_type;
 }
 
-void zend_interned_strings_dtor(void)
+ZEND_API void zend_interned_strings_dtor(zend_interned_strings_init_stage storage_type)
 {
-#ifndef ZTS
-	zend_hash_destroy(&CG(interned_strings));
-#else
-	uint32_t i;
-
-	for (i = 0; i < CG(known_strings_count); i++) {
-		zend_zts_interned_string_free(&CG(known_strings)[i]);
+	if (ZEND_INTERNED_STRINGS_UNINITIALIZED == storage_type || storage_type != zend_strings_storage_type) {
+		return;
 	}
-#endif
-	free(CG(known_strings));
-	CG(known_strings) = NULL;
-	CG(known_strings_count) = 0;
+
+	zend_hash_destroy(&interned_strings_permanent);
+	zend_hash_destroy(&interned_strings_volatile);
+
+	free(known_interned_strings);
 	known_interned_strings = NULL;
 	known_interned_strings_count = 0;
+#ifndef ZTS
+	CG(known_strings) = NULL;
+	CG(known_strings_count) = 0;
+#endif
+
+	zend_strings_storage_type = ZEND_INTERNED_STRINGS_UNINITIALIZED;
 }
 
-static zend_string *zend_new_interned_string_int(zend_string *str)
+static zend_string *zend_interned_string_ht_lookup(zend_string *str, HashTable *interned_strings)
 {
-#ifndef ZTS
 	zend_ulong h;
 	uint32_t nIndex;
 	uint32_t idx;
 	Bucket *p;
 
-	if (ZSTR_IS_INTERNED(str)) {
-		return str;
+	/* Handle the passed string as read only. */
+	h = ZSTR_H(str);
+	if (!h) {
```

Comment:
Indeed. I was worried about a possible race condition, but zend_string_forget_hash_val() is currently used in thread safe manner, and otherwise an interned string will always have a hash val. Fixing.